src/v/cloud_storage/remote_partition.cc

```diff
@@ -342,6 +355,8 @@ class partition_record_batch_reader_impl final
                 try {
                     auto result = co_await _reader->read_some(
                       deadline, *_ot_state);
+                    throw_on_external_abort();
```

Comment:
Wouldn't we really want to pass the abort source further down to `read_some`? Otherwise if we get stuck on any of the reads we will never reach this?

Not saying this isn't an improvement but more interested in a general discussion as I am currently looking into a different case where we "leak" background futures that are stuck on reads/writes under load. It feels like we need more abort_sources in general and also need to make sure we pass them all the way down.

Reply:
That makes sense to me, I had assumed the `remote_segment_batch_consumer` took note of it's `abort_source`, but it does not appear to.

Maybe something like this:
```diff
diff --git a/src/v/cloud_storage/remote_segment.cc b/src/v/cloud_storage/remote_segment.cc
index 87b3d9939b..c9e3fec90b 100644
--- a/src/v/cloud_storage/remote_segment.cc
+++ b/src/v/cloud_storage/remote_segment.cc
@@ -1270,6 +1270,12 @@ public:
             co_return stop_parser::yes;
         }
 
+        if (
+          _config.abort_source.has_value()
+          && _config.abort_source.value().get().abort_requested()) {
+            co_return stop_parser::yes;
+        }
+
         co_return stop_parser::no;
     }
```

Reply:
This PR adds a connection-level abort_source, it's possible that a request-level abort_source would also be useful, which could chain onto the connection-level one.

Reply:
Most of the time the abort_source will be triggered when the underlying `remote_segment_batch_readfer` will be waiting inside the `remote_segment::hydrate` call and it won't be able to quit early. The latency there is in the ballpark of 1s. So I agree that it should be handled one level below.

Reply:
I think it's actually desirable to finish the hydration even if the request got aborted. In all likelihood, the client will retry, so if we cancel the hydration we'd just be throwing away work.

Reply:
I totally agree. The hydration should continue. But inside the `read_some` call there is a call to `hydrate` which will block the fiber from being aborted. I'm not sure moving the reader to the evicted list will solve this. The reader will move but the fiber that actually reads data for the fetch request will be stuck until the segment is downloaded. The `hydrate` method itself is actually just waiting on a future. The hydration is done in the background so I think it should be possible to abort it without canceling hydration. The background hydration loop should be prepared to this and handle `broken_promise` exception.

Reply:
Maybe for this PR it's fine. But we should consider propagating abort source further.